CMakeLists.txt

```diff
@@ -16,6 +16,11 @@ ament_target_dependencies(${PROJECT_NAME} "rclcpp" "rclcpp_action")
 target_link_libraries(${PROJECT_NAME} ${CMAKE_THREAD_LIBS_INIT})
 target_include_directories(${PROJECT_NAME} PUBLIC include)
 
+# A library to detect a realtime kernel and set thread priority, if one is found
+add_library(thread_priority SHARED src/thread_priority.cpp)
+target_include_directories(thread_priority PRIVATE include)
```

Comment:
This is probably wrong. Have you tried using this somewhere?

Reply:
Yep, this PR from ros2_control:  https://github.com/ros-controls/ros2_control/pull/794

Reply:
I don't understand why this would work because PRIVATE means that these includes are not part of your exported interface, when clearly they are.

Reply:
I think normally this should be:
```cmake
target_include_directories(thread_priority
  $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>
  $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}>)
```

Reply:
PRIVATE here makes sense in the ros2_control plugins that are not libraries because they don't export headers as they are just plugins.

Reply:
Your suggestion -->

CMake Error at CMakeLists.txt:21 (target_include_directories):
  target_include_directories called with invalid arguments

I'm looking into it

Reply:
This works:

```
target_include_directories(thread_priority PUBLIC
  "$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>"
  "$<INSTALL_INTERFACE:include>"
)
```

Reply:
I think this should be as it was, but with "PUBLIC". I was also using this complex definitions few times, but never found out where they actually have to be used. I mean why make CMakeLists.txt less readable if there is no advantage of it.

Reply:
That generator expression ensures that when building the library you get the proper public include directory but it also ensures that when you install the library, the installed library doesn't include `/home/my_user/Repos/realtime_tools/include` since such a user-specific include directory makes no sense for shipping libraries and is going to break even on your own machine should you ever delete the source code repo from which you built the library.

Yes, the more complicated form looks very tedious but the simple `target_include_directories(libName PUBLIC include)` form is insufficient for library targets that need to get installed.

Reply:
I have different experience. We are using simpler form in whole ros2_control, which are basically many libraries and people don't have any problems executing them. I understand, but can not confirm you explanation from my experience. Especially, look at few lines above that we use simple from which is used in *every* controller in `ros2_controllers` repository and we have never had any issues you described.

Is this because we are using ROS build farm?

Reply:
If I interpret what @ChrisThrasher wrote correctly, you wouldn't see any broken functionality from `target_include_directories(libName PUBLIC include)`, but you might get some extra junk installed. Only if you have already built from source on the same machine, though.

I trust him a lot, he's the PickNik expert on CMake so I won't change this.